_posts/2025-04-22-test-classloading-rewrite.adoc

```diff
@@ -0,0 +1,63 @@
+---
+layout: post
+title: 'The internals (and a few externals) of Quarkus test classloading have changed'
+tags: announcement testing
+synopsis: 'The way that Quarkus loads test classes has been updated. Most tests will not need to change, but here are some things to watch out for.'
+author: hcummins
+---
+
+== What's changing?
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
+Other test extensions haven't changed, nor has dev mode and production mode classloading.
+This change should make Quarkus testing work better, and it allowed us to fix a pile of longstanding bugs.
+However, it did introduce a few bugs we know about, and most likely also some bugs we don't yet know about.
+We're keen to get feedback from the community so that we can get fixing.
+
+== Why?
+
+In previous versions, Quarkus tests were invoked using the default JUnit classloader, and then executed in a different, Quarkus-aware, classloader.
+
+This mostly worked very well, and meant that `QuarkusTest` tests mostly behaved as if they were part of the same application as the code under test.
+The Quarkus test framework could start and stop Quarkus instances at the right point in the test lifecycle, inject CDI dependencies, and do other useful Quarkus bytecode manipulation.
```

Comment:
point or time?

Reply:
I'd say point, because we don't judge it by a clock, we do it relative to other events in the lifecycle. But I don't have strong feelings, if you think time is clearer?

Reply:
point is fine - I was just wondering.